File: foobarqix/src/main/java/FooBarQix.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class FooBarQix {

    private static Map<Integer, String> mapIntString = new HashMap<>();

    static {
        mapIntString.put(0, "*");
        mapIntString.put(3, "Foo");
        mapIntString.put(5, "Bar");
        mapIntString.put(7, "Qix");
    }

    public String compute(String numberString) {
        boolean isNumber = true;
        String returnValue = "";
        final Integer numberInt;

        try {
             numberInt = Integer.valueOf(numberString);
        } catch (NumberFormatException exception) {
            return "";
        }

        // Handles divisible by
        List<Integer> numbersToCheck = Arrays.asList(3, 5, 7);
        for (Integer element: numbersToCheck) {
            if (numberInt % element == 0) {
                returnValue += mapIntString.get(element);
                isNumber = false;
            }
        }

        // Handles the contains
        for (char element: numberString.toCharArray()) {
            if (numbersToCheck.contains(Character.getNumericValue(element))) {
                returnValue += mapIntString.get(Character.getNumericValue(element));
                isNumber = false;
            }
            // Handles the 0 to add if it contains Foo or Bar or Qix
            if (!isNumber && Character.getNumericValue(element) == 0) {
                returnValue += mapIntString.get(Character.getNumericValue(element));
            }
        }

        returnValue = returnValue.isEmpty() ? numberInt.toString() : returnValue;

        // Handles the 0 to add if it is still a number
        return returnValue.replace("0", mapIntString.get(0));

    }

}

File: foobarqix/src/main/kotlin/FooBarQix.kt
val mapIntString = mapOf(0 to "*", 3 to "Foo", 5 to "Bar", 7 to "Qix")
val replaceNumberList = listOf("Foo", "Bar", "Qix")

fun Char.toIntValue() = Character.getNumericValue(this)

fun compute(numberString: String): String {
    val numberInt = numberString.toIntOrNull() ?: return ""
    var returnValue = ""

    val numbersToCheck = listOf(3, 5, 7)
    numbersToCheck.asSequence()
        .filter { numberInt % it == 0 }
        .forEach { returnValue += mapIntString[it]}

    numberString.asSequence()
        .filter { mapIntString.containsKey(it.toIntValue()) }
        .forEach { returnValue += mapIntString[it.toIntValue()]}

    val keepString = replaceNumberList.any { returnValue.contains(it) }

    return if (keepString) returnValue else numberString.replace("0", mapIntString[0] ?: "0")
}